### PR TITLE
mcs: close the connection to avoid retrying to connect

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #5839.

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Manual test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -1332,6 +1331,8 @@ func (s *GrpcServer) RegionHeartbeat(stream pdpb.PD_RegionHeartbeatServer) error
 				if cancel != nil {
 					cancel()
 				}
+
```
Comment (lhy1024) [inline server/grpc_service.go:1334]: if only switch primary, do we need to close it?

File: server/forward.go [hunk 2]
```diff
@@ -363,25 +372,12 @@ func (s *GrpcServer) getDelegateClient(ctx context.Context, forwardedHost string
...
-			switch serviceName[0] {
-			case utils.TSOServiceName:
-				return "", ErrNotFoundTSOAddr
-			case utils.SchedulingServiceName:
-				return "", ErrNotFoundSchedulingAddr
-			}
-		}
-	} else if fh := grpcutil.GetForwardedHost(streamCtx); !s.isLocalRequest(fh) {
-		forwardedHost = fh
+func (s *GrpcServer) closeDelegateClient(forwardedHost string) {
```
Comment (lhy1024) [inline server/forward.go:375]: Do we need to add some log?

**Comment by lhy1024** [comment]
image
It works too.

File: server/grpc_service.go (hunk 1 above)
Comment (rleungx) [reply server/grpc_service.go:1334]: Both are ok to me.

**Review by CabinfeverB — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

File: pkg/utils/grpcutil/grpcutil.go [hunk 3]
```diff
@@ -252,6 +252,9 @@ func CheckStream(ctx context.Context, cancel context.CancelFunc, done chan struc
 
 // NeedRebuildConnection checks if the error is a connection error.
 func NeedRebuildConnection(err error) bool {
+	if err == nil {
+		return false
+	}
 	return err == io.EOF ||
```
Comment (JmPotato) [inline pkg/utils/grpcutil/grpcutil.go:258]:
```suggestion
	return err != nil && err == io.EOF ||
```

File: server/forward.go [hunk 4]
```diff
@@ -131,26 +135,31 @@ func (s *GrpcServer) forwardTSO(stream pdpb.PD_TsoServer) error {
...
+				tsoStreamErr = errors.WithStack(err)
+				return tsoStreamErr
 			}
 			lastForwardedHost = forwardedHost
 		}
 
 		tsopbResp, err := s.forwardTSORequestWithDeadLine(forwardCtx, cancelForward, forwardStream, request, tsDeadlineCh)
 		if err != nil {
+			tsoStreamErr = errors.WithStack(err)
 			return errors.WithStack(err)
```
Comment (JmPotato) [inline server/forward.go:162]: Should this be `return tsoStreamErr`?

Comment (rleungx) [reply server/forward.go]: fixed

**Review by JmPotato — APPROVED** [review]

**Pull request merged**